src/main/java/com/paloaltonetworks/osc/api/PANManagerSecurityGroupInterfaceApi.java

```diff
@@ -1,87 +1,171 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ *    License for the specific language governing permissions and limitations
+ *    under the License.
+ */
+package com.paloaltonetworks.osc.api;
+
+import static com.paloaltonetworks.utils.TagToSGIdUtil.securityGroupTag;
+import static com.paloaltonetworks.utils.VsToDevGroupNameUtil.devGroupName;
+import static java.util.Collections.emptySet;
+import static java.util.stream.Collectors.toSet;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.osc.sdk.manager.api.ManagerSecurityGroupInterfaceApi;
+import org.osc.sdk.manager.element.ApplianceManagerConnectorElement;
+import org.osc.sdk.manager.element.ManagerSecurityGroupInterfaceElement;
+import org.osc.sdk.manager.element.SecurityGroupInterfaceElement;
+import org.osc.sdk.manager.element.VirtualSystemElement;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.paloaltonetworks.panorama.api.mapping.AddressEntry;
+import com.paloaltonetworks.panorama.api.methods.PanoramaApiClient;
+
+/**
+ * This documents "Device Management Apis"
+ */
+public class PANManagerSecurityGroupInterfaceApi implements ManagerSecurityGroupInterfaceApi {
+
+    private static final Logger LOG = LoggerFactory.getLogger(PANManagerSecurityGroupInterfaceApi.class);
+
+    private static final Character IDSTRING_SEPARATOR = '_';
+
+    private VirtualSystemElement vs;
+    private ApplianceManagerConnectorElement mc;
+    private PanoramaApiClient panClient;
+
+    private String devGroup;
+
+    public PANManagerSecurityGroupInterfaceApi(ApplianceManagerConnectorElement mc, VirtualSystemElement vs,
+            PanoramaApiClient panClient) {
+        this.vs = vs;
+        this.mc = mc;
+        this.panClient = panClient;
+        this.devGroup = devGroupName(vs.getId());
+    }
+
+    @Override
+    public String createSecurityGroupInterface(SecurityGroupInterfaceElement sgiElement) throws Exception {
+        return doUpdateSecurityGroupInterface(sgiElement);
+    }
+
+    @Override
+    public void updateSecurityGroupInterface(SecurityGroupInterfaceElement sgiElement) throws Exception {
+        doUpdateSecurityGroupInterface(sgiElement);
+    }
+
+    @Override
+    public void deleteSecurityGroupInterface(String sgInterfaceId) throws Exception {
+        String sgId = extractSGId(sgInterfaceId);
+        String sgTag = securityGroupTag(sgId);
+
+        List<AddressEntry> ipsOnMgr = this.panClient.fetchAddressesWithTag(sgTag, this.devGroup);
+
+        if (ipsOnMgr.isEmpty()) {
+            LOG.info("Security group {} on VS {} has no addresses. Delete SGi is a no-op.", sgId, this.vs.getId());
+        }
+
+        for (AddressEntry entry : ipsOnMgr) {
+            List<String> tags = entry.getTagNames();
+            tags.remove(sgTag); // only policy tags left!

```

Comment:
is this removing the "web", production tags or the OSC_SecurityGroup_ tags?

my understanding is we should remove the web/production tags basically the tags returned by org.osc.sdk.manager.element.ManagerSecurityGroupInterfaceElement.getManagerPolicyElements()